### PR TITLE
[Admin] homepage statistic

### PR DESCRIPTION
**Pull request by ghost**
## Description
- Statistic Admin Homepage
## Migrations
NO

## Impacted Areas in Application
List general components of the application that this PR will affect:

- routes/api.php

## SQL query and `explain`

1. explain select count(*) as aggregate from `users` where `created_at` > '2018-04-29 02:07:16'\G

> *************************** 1. row ***************************
>            id: 1
>   select_type: SIMPLE
>         table: users
>    partitions: NULL
>          type: ALL
> possible_keys: NULL
>           key: NULL
>       key_len: NULL
>           ref: NULL
>          rows: 10
>      filtered: 33.33
>         Extra: Using where
> 1 row in set, 1 warning (0.00 sec)

2. explain select count(*) as aggregate from `posts` where `created_at` > '2018-04-29 02:07:16' and `posts`.`deleted_at` is null\G

> *************************** 1. row ***************************
>            id: 1
>   select_type: SIMPLE
>         table: posts
>    partitions: NULL
>          type: ALL
> possible_keys: NULL
>           key: NULL
>       key_len: NULL
>           ref: NULL
>          rows: 50
>      filtered: 3.33
>         Extra: Using where
> 1 row in set, 1 warning (0.00 sec)

3. explain select count(*) as aggregate from `orders` where `created_at` > '2018-04-29 02:07:16' and `orders`.`deleted_at` is null\G

> *************************** 1. row ***************************
>            id: 1
>   select_type: SIMPLE
>         table: orders
>    partitions: NULL
>          type: ALL
> possible_keys: NULL
>           key: NULL
>       key_len: NULL
>           ref: NULL
>          rows: 10
>      filtered: 10.00
>         Extra: Using where
> 1 row in set, 1 warning (0.00 sec)

4. explain select sum(`quantity`) as aggregate from `order_details` where exists (select * from `orders` where `order_details`.`order_id` = `orders`.`id` and `status` = '1' and `orders`.`deleted_at` is null) and `order_details`.`deleted_at` is null\G

> *************************** 1. row ***************************
>            id: 1
>   select_type: PRIMARY
>         table: order_details
>    partitions: NULL
>          type: ALL
> possible_keys: NULL
>           key: NULL
>       key_len: NULL
>           ref: NULL
>          rows: 10
>      filtered: 10.00
>         Extra: Using where
> *************************** 2. row ***************************
>            id: 2
>   select_type: DEPENDENT SUBQUERY
>         table: orders
>    partitions: NULL
>          type: eq_ref
> possible_keys: PRIMARY
>           key: PRIMARY
>       key_len: 4
>           ref: product_tiki.order_details.order_id
>          rows: 1
>      filtered: 10.00
>         Extra: Using where
> 2 rows in set, 2 warnings (0.01 sec)

5. explain select * from `products` where `created_at` > '2018-04-29 02:07:16' and `products`.`deleted_at` is null order by `avg_rating` desc limit 5\G

> *************************** 1. row ***************************
>            id: 1
>   select_type: SIMPLE
>         table: products
>    partitions: NULL
>          type: ALL
> possible_keys: NULL
>           key: NULL
>       key_len: NULL
>           ref: NULL
>          rows: 20
>      filtered: 5.00
>         Extra: Using where; Using filesort
> 1 row in set, 1 warning (0.00 sec)

6. explain select * from `categories` where `categories`.`id` in ('7', '12', '17', '26')\G

> *************************** 1. row ***************************
>            id: 1
>   select_type: SIMPLE
>         table: categories
>    partitions: NULL
>          type: range
> possible_keys: PRIMARY
>           key: PRIMARY
>       key_len: 4
>           ref: NULL
>          rows: 4
>      filtered: 100.00
>         Extra: Using where
> 1 row in set, 1 warning (0.00 sec)

7. explain select `users`.*, (select count(*) from `comments` where `users`.`id` = `comments`.`user_id` and `comments`.`deleted_at` is null) as `comments_count`, (select count(*) from `posts` where `users`.`id` = `posts`.`user_id` and `posts`.`deleted_at` is null) as `posts_count` from `users`\G

> *************************** 1. row ***************************
>            id: 1
>   select_type: PRIMARY
>         table: users
>    partitions: NULL
>          type: ALL
> possible_keys: NULL
>           key: NULL
>       key_len: NULL
>           ref: NULL
>          rows: 10
>      filtered: 100.00
>         Extra: NULL
> *************************** 2. row ***************************
>            id: 3
>   select_type: DEPENDENT SUBQUERY
>         table: posts
>    partitions: NULL
>          type: ref
> possible_keys: posts_user_id_foreign
>           key: posts_user_id_foreign
>       key_len: 4
>           ref: product_tiki.users.id
>          rows: 5
>      filtered: 10.00
>         Extra: Using where
> *************************** 3. row ***************************
>            id: 2
>   select_type: DEPENDENT SUBQUERY
>         table: comments
>    partitions: NULL
>          type: ref
> possible_keys: comments_user_id_foreign
>           key: comments_user_id_foreign
>       key_len: 4
>           ref: product_tiki.users.id
>          rows: 5
>      filtered: 10.00
>         Extra: Using where
> 3 rows in set, 3 warnings (0.00 sec)

8. explain select * from `user_info` where `user_info`.`user_id` in ('1', '2', '3', '4', '5', '6', '7', '8', '9', '10')\G

> *************************** 1. row ***************************
>            id: 1
>   select_type: SIMPLE
>         table: user_info
>    partitions: NULL
>          type: ALL
> possible_keys: user_info_user_id_foreign
>           key: NULL
>       key_len: NULL
>           ref: NULL
>          rows: 10
>      filtered: 100.00
>         Extra: Using where
> 1 row in set, 1 warning (0.00 sec)

9. explain select `orders`.*, (select count(*) from `order_details` where `orders`.`id` = `order_details`.`order_id` and `order_details`.`deleted_at` is null) as `orderdetails_count` from `orders` where `orders`.`deleted_at` is null order by `total` desc limit 5\G

> *************************** 1. row ***************************
>            id: 1
>   select_type: PRIMARY
>         table: orders
>    partitions: NULL
>          type: ALL
> possible_keys: NULL
>           key: NULL
>       key_len: NULL
>           ref: NULL
>          rows: 10
>      filtered: 10.00
>         Extra: Using where; Using filesort
> *************************** 2. row ***************************
>            id: 2
>   select_type: DEPENDENT SUBQUERY
>         table: order_details
>    partitions: NULL
>          type: ref
> possible_keys: order_details_order_id_foreign
>           key: order_details_order_id_foreign
>       key_len: 4
>           ref: product_tiki.orders.id
>          rows: 1
>      filtered: 10.00
>         Extra: Using where
> 2 rows in set, 2 warnings (0.00 sec)

10. explain select * from `users` where `users`.`id` in ('2', '5', '8')\G

> *************************** 1. row ***************************
>            id: 1
>   select_type: SIMPLE
>         table: users
>    partitions: NULL
>          type: range
> possible_keys: PRIMARY
>           key: PRIMARY
>       key_len: 4
>           ref: NULL
>          rows: 3
>      filtered: 100.00
>         Extra: Using where
> 1 row in set, 1 warning (0.01 sec)

### CODE REVIEW

File: public/js/homepage.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    });
+    $( '.dropdown-order li a' ).click( function (event) {
+        event.preventDefault();
+        url = $(this).attr("href");
+        $.ajax({
+            method: 'GET',
+            url: url,
+            success: function (data) {
+                $('#order_time').text(data.time);
+                for (let i = 0; i < 5; i++) {
```
Comment (at-xinhnguyen) [inline public/js/homepage.js]: sao em dùng hard ở đây vây

Comment (at-phuongvu) [reply public/js/homepage.js]: dạ vì API sẽ trả về 5 record ạ :3 nên em trỏ luôn ạ. Vì logic chỗ đó là select top 5 nên em thấy ko cần phải lấy `length` của `data['user']` hay `data['order']`

Comment (at-xinhnguyen) [reply public/js/homepage.js]: em config chung 1 chỗ đi. chứ lỡ sau này muốn `top4` thì chỗ này lỗi sao em

File: app/Http/Controllers/Api/HomeController.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    *
+    * @return \Illuminate\Http\Response
+    */
+    public function index(Request $request)
+    {
+        $timeStr = 'last ';
+        if ($request->time) {
+            $timeStr .= $request->time;
+        } else {
+            $timeStr = 'first day of January 2001';
```
Comment (at-xinhnguyen) [inline app/Http/Controllers/Api/HomeController.php]: sao lại hard code vậy em

File: app/Http/Controllers/Api/HomeController.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            }, 'posts' => function ($query) use ($time) {
+                return $query->where('created_at', '>', $time);
+            }])->get();
+
+            foreach ($users as $user) {
+                $pointCalculated = $user->comments_count + $user->posts_count;
+                $user['point'] = $pointCalculated;
+                $user['routes'] = route('admin.users.show', array('id' => $user->id));
+            }
+            $data['users'] = array_values($users->sortByDesc('point')->take(5)->toarray());
```
Comment (at-xinhnguyen) [inline app/Http/Controllers/Api/HomeController.php]: em config số 5 vào với sửa `toarray` thành `toArray`

File: app/Http/Controllers/Api/HomeController.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+        if ($request->time) {
+            $timeStr = 'last '.$request->time;
+            $time = new Carbon($timeStr);
+        } else {
+            $time = new Carbon(Carbon::minValue());
+            $request->time = 'all';
+        }
+
+        if ($request->type == 'user') {
```
Comment (at-huenguyen) [inline app/Http/Controllers/Api/HomeController.php]: `'user'` cho vào config nha em

File: app/Http/Controllers/Api/HomeController.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            $users = User::with('userInfo')->withCount(['comments' => function ($query) use ($time) {
+                return $query->where('created_at', '>', $time);
+            }, 'posts' => function ($query) use ($time) {
+                return $query->where('created_at', '>', $time);
+            }])->get();
+
+            foreach ($users as $user) {
+                $pointCalculated = $user->comments_count + $user->posts_count;
+                $user['point'] = $pointCalculated;
```
Comment (at-huenguyen) [inline app/Http/Controllers/Api/HomeController.php]: chỗ này em muốn lấy giá trị `point` của thèn cuối cùng thôi à?

File: app/Http/Controllers/Api/HomeController.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+                $user['routes'] = route('admin.users.show', array('id' => $user->id));
+            }
+            $data['users'] = array_values($users->sortByDesc('point')->take(config('define.homepage.numberOfRecords'))->toArray());
+        }
+
+        if ($request->type == 'order') {
```
Comment (at-huenguyen) [inline app/Http/Controllers/Api/HomeController.php]: `'order'` bỏ config nha em

File: app/Http/Controllers/Api/HomeController.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    * Display a listing of the resource.
+    *
+    * @param Illuminate\Http\Request $request time range and type
+    *
+    * @return \Illuminate\Http\Response
+    */
+    public function index(Request $request)
+    {
+
+        if ($request->time) {
```
Comment (at-huenguyen) [inline app/Http/Controllers/Api/HomeController.php]: đoạn `if..else` này là sao em? chị chưa rõ chỗ này

File: app/Http/Controllers/Api/HomeController.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            $users = User::with('userInfo')->withCount(['comments' => function ($query) use ($time) {
+                return $query->where('created_at', '>', $time);
+            }, 'posts' => function ($query) use ($time) {
+                return $query->where('created_at', '>', $time);
+            }])->get();
+
+            foreach ($users as $user) {
+                $pointCalculated = $user->comments_count + $user->posts_count;
+                $user['point'] = $pointCalculated;
+                $user['routes'] = route('admin.users.show', array('id' => $user->id));
```
Comment (at-huenguyen) [inline app/Http/Controllers/Api/HomeController.php]: chỗ này là sao em?

**Review by ghost — CHANGES_REQUESTED** [review]
plz check my cmts

File: app/Http/Controllers/Api/HomeController.php (hunk 8 above)
Comment (at-phuongvu) [reply app/Http/Controllers/Api/HomeController.php]: chỗ này em trả về `route` để update lại đường link ạ

File: app/Http/Controllers/Api/HomeController.php (hunk 7 above)
Comment (at-phuongvu) [reply app/Http/Controllers/Api/HomeController.php]: dạ chỗ này em đưa `time=month` hoặc `time=year` lên đường link rồi em lấy cái chữ `month` `year` đó để ghép vào chữ `last` rồi tạo ra `Carbon('last ....')` còn nếu người dùng chọn all time thì em tạo ra Carbon minValue() để so sánh luôn đúng ạ

File: app/Http/Controllers/Api/HomeController.php (hunk 5 above)
Comment (at-phuongvu) [reply app/Http/Controllers/Api/HomeController.php]: mỗi thằng `$user` sẽ có 1 giá trị `point` riêng vì em lấy mảng bên trong mà ạ

File: resources/views/admin/pages/index.blade.php [hunk 9]
```diff
@@ -1,37 +1,43 @@
...
+  <link rel="stylesheet" href="/css/homepage.css">
+@endsection
 @section('content')
+
 <div class="right_col" role="main">
   <div class="">
     <div class="row top_tiles">
       <div class="x_title col-md-12">
-        <h2>{{ __('homepage.admin.monthly_statistic') }}</h2>
+        <a class="btn col-md-3 @if (app('request')->input('time') == null || app('request')->input('time') == 'month') active-filter @endif" href="{{ route('admin.home', ['time' => 'month']) }}">{{ __('homepage.admin.statistic.monthly') }}</a>
```
[the remark this answers is not in the document]
Comment (at-phuongvu) [reply resources/views/admin/pages/index.blade.php]: chị ơi em thêm vào `config` mà viết ` 'month' => 'month'` thì có vô nghĩa quá không ạ :(

[the remark this answers is not in the document]
Comment (at-huenguyen) [reply resources/views/admin/pages/index.blade.php]: @at-phuongvu ko đâu em, bữa sau gỉa sử biến `month` bị thay đổi sao em

File: app/Http/Controllers/Api/HomeController.php (hunk 5 above)
Comment (at-huenguyen) [reply app/Http/Controllers/Api/HomeController.php]: lấy mảng thì dùng `$user['point'][] = ...` chứ hè.
để vại nó có đc mảng ko em

File: app/Http/Controllers/Api/HomeController.php (hunk 8 above)
Comment (at-huenguyen) [reply app/Http/Controllers/Api/HomeController.php]: mà cái này là em viết cho `API` mà, nó phản hồi lại 1 mảng dữ liệu gì đó chứ

Comment (at-phuongvu) [reply app/Http/Controllers/Api/HomeController.php]: dạ đúng ạ ... nó trả về string route thôi rồi em gán lại cho cái `href` ạ ... vì muốn cập nhật đường link mà ạ :D

File: app/Http/Controllers/Api/HomeController.php (hunk 5 above)
Comment (at-phuongvu) [reply app/Http/Controllers/Api/HomeController.php]: dạ được ạ :D em gán cái key `point` cho `user` chứ không phải `users` ạ nên mỗi `user` sẽ nhận giá trị của nó

**Review by ghost — CHANGES_REQUESTED** [review]

File: resources/views/admin/pages/index.blade.php [hunk 10]
```diff
@@ -1,37 +1,43 @@
...
+  <link rel="stylesheet" href="/css/homepage.css">
+@endsection
 @section('content')
+
 <div class="right_col" role="main">
   <div class="">
     <div class="row top_tiles">
       <div class="x_title col-md-12">
-        <h2>{{ __('homepage.admin.monthly_statistic') }}</h2>
+        <a class="btn col-md-3 @if (app('request')->input('time') == config('define.homepage.time.month') || !app('request')->input('time')) active-filter @endif" href="{{ route('admin.home', ['time' => 'month']) }}">{{ __('homepage.admin.statistic.monthly') }}</a>
```
[the remark this answers is not in the document]
Comment (at-phuongvu) [reply resources/views/admin/pages/index.blade.php]: week year em làm hết rồi mà ạ ... trong config em trỏ `'time' => [ ]` mà ạ

**Review by ghost — APPROVED** [review]
LGTM

**Pull request merged**